python/python/bystro/covariance/_base_covariance.py

```diff
@@ -0,0 +1,549 @@
+
+    Parameters
+    ----------
+    covariance : np.array-like(p,p)
+        The covariance matrix
+
+    Returns
+    -------
+    srank : float
+        The stable rank
+    """
+    singular_values = la.svd(covariance, compute_uv=False)
+    srank = np.sum(singular_values) / singular_values[0]
+    return srank
+
+
+def _predict(covariance, Xobs, idxs):
+    """
+    Predicts missing data using observed data.
+
+    Parameters
+    ----------
+    covariance : np.array-like(p,p)
+        The covariance matrix
+
+    Xobs : np array-like,(N_samples,\sum idxs)
+        The observed data
+
+    idxs: np.array-like,(sum(p),)
+        The observation locations
+
+    Returns
+    -------
+    preds : np.array-like,(N_samples,p-\sum idxs)
+        The predicted values
+    """
+    covariance_sub = covariance[idxs == 1]
+    covariance_22 = covariance_sub[:, idxs == 1]
+    covariance_21 = covariance_sub[:, idxs == 0]
+
+    covariance_nonsub = covariance[idxs == 0]
+    covariance_11 = covariance_nonsub[:, idxs == 0]
+
+    beta_bar = la.solve(covariance_22, covariance_21)
+    Second_part = np.dot(covariance_21.T, beta_bar)
+
+    covariance_11 - Second_part
+
+    preds = np.dot(Xobs[:, idxs == 1], beta_bar)
+    return preds
+
+
+#########################################
+#########################################
+##                                     ##
+##  Gaussian Likelihood-based methods  ##
+##                                     ##
+#########################################
+#########################################
+
+
+def _conditional_score(covariance, X, idxs, weights=None):
+    """
+    Returns the predictive log-likelihood of a subset of data.
+
+    mean(log p(X[idx==1]|X[idx==0],covariance))
+
+    Parameters
+    ----------
+    covariance : np.array-like(p,p)
+        The covariance matrix
+
+    X : np.array-like,(N,sum(idxs))
+        The data
+
+    idxs: np.array-like,(sum(p),)
+        The observation locations
+
+    weights : np.array-like,(N,),default=None
+        The optional weights on the samples
+
+    Returns
+    -------
+    avg_score : float
+        Average log likelihood
+    """
+    if weights is None:
+        weights = np.ones(X.shape[0])
+    avg_score = np.mean(
+        weights * _conditional_score_samples(covariance, X, idxs)
+    )
+    return avg_score
+
+
+def _conditional_score_samples(covariance, X, idxs):
+    """
+    Return the conditional log likelihood of each sample, that is
+
+    log p(X[idx==1]|X[idx==0],covariance) = N(Sigma_10Sigma_00^{-1}x_0,
+                                    Sigma_11-Sigma_10Sigma_00^{-1}Sigma_01)
+
+    Parameters
+    ----------
+    covariance : np.array-like(p,p)
+        The covariance matrix
+
+    X : np.array-like,(N,p)
+        The data
+
+    idxs: np.array-like,(p,)
+        The observation locations
+
+    Returns
+    -------
+    scores : float
+        Log likelihood for each sample
+    """
+    covariance_sub = covariance[idxs == 1]
```

Comment:
should this block be refactored in light of the `_predict` method above?

Reply:
No,  pred method just returns means, not covariance which is needed for likelihood.

Reply:
Here are six different places in the code where we're repeating the same pattern line for line:

```python

    covariance_sub = covariance[idxs == 1]
    covariance_22 = covariance_sub[:, idxs == 1]
    covariance_21 = covariance_sub[:, idxs == 0]
    beta_bar = la.solve(covariance_22, covariance_21)
###
    covariance_sub = covariance[idxs == 1]
    covariance_22 = covariance_sub[:, idxs == 1]
    covariance_21 = covariance_sub[:, idxs == 0]
    beta_bar = la.solve(covariance_22, covariance_21)

###

    covariance_sub = covariance[idxs == 1]
    covariance_22 = covariance_sub[:, idxs == 1]
    covariance_21 = covariance_sub[:, idxs == 0]
    beta_bar = la.solve(covariance_22, covariance_21)

####
    csub = cov[idxs == 1]
    c22 = csub[:, idxs == 1]
    c21 = csub[:, idxs == 0]
    csub = cov[idxs == 0]
    beta_bar = la.solve(c22, c21)

####
    csub = cov[idxs == 1]
    c22 = csub[:, idxs == 1]
    c21 = csub[:, idxs == 0]
    beta_bar = la.solve(c22, c21)


####
    csub = cov[idxs == 1]
    c22 = csub[:, idxs == 1]
    c21 = csub[:, idxs == 0]
    beta_bar = la.solve(c22, c21)

```

My concerns are that:

1. it will be easy for one of these instances of the pattern to get out of synch with the others, and once it gets out of synch it will be difficult to tell that it's out of synch precisely because it's repeated so many times.

2.  A pattern repeated that many times probably has a pre-existing name.  If we pull it out into a helper method and call it by its proper name, we can help the reader out.  I don't think it's quite self-evident to the modal engineer what the intent is here (using `la.solve` in order to compute a log normal density isn't really elementary), so a named function would give them something to google.